manager/integration/tests/test_scheduling.py

```diff
@@ -0,0 +1,349 @@
+    :param size: The size of the volume, as a string representing the number
+    of bytes.
+    :return: The volume instance created.
+    """
+    client.create_volume(name=volume_name, size=size,
+                         numberOfReplicas=num_of_replicas)
+    volume = wait_for_volume_detached(client, volume_name)
+    assert volume["name"] == volume_name
+    assert volume["size"] == size
+    assert volume["numberOfReplicas"] == num_of_replicas
+    assert volume["created"] != ""
+    return volume
+
+
+def cleanup(client, volume):  # NOQA
+    """
+    Clean up the volume after the test.
+    :param client: The Longhorn client to use in the request.
+    :param volume: The volume to clean up.
+    """
+    volume.detach()
+    volume = wait_for_volume_detached(client, volume["name"])
+    client.delete(volume)
+    wait_for_volume_delete(client, volume["name"])
+    volumes = client.list_volume()
+    assert len(volumes) == 0
+
+
+def get_host_replica(volume, host_id):
+    """
+    Get the replica of the volume that is running on the test host. Trigger a
+    failed assertion if it can't be found.
+    :param volume: The volume to get the replica from.
+    :param host_id: The ID of the test host.
+    :return: The replica hosted on the test host.
+    """
+    host_replica = None
+    for i in volume["replicas"]:
+        if i["hostId"] == host_id:
+            host_replica = i
+    assert host_replica is not None
+    return host_replica
+
+
+# We check to make sure the replica is found, running, and in RW mode (not
+# rebuilding) since the longhorn-engine has the latest status compared to
+# longhorn-manager, which might be in an intermediate state.
+def wait_new_replica_ready(client, volume_name, replica_names):  # NOQA
+    """
+    Wait for a new replica to be found on the specified volume. Trigger a
+    failed assertion if one can't be found.
+    :param client: The Longhorn client to use in the request.
+    :param volume_name: The name of the volume.
+    :param replica_names: The list of names of the volume's old replicas.
+    """
+    new_replica_ready = False
+    for _ in range(RETRY_COUNTS):
+        v = client.by_id_volume(volume_name)
+        for r in v["replicas"]:
+            if r["name"] not in replica_names and r["running"] and \
+                    r["mode"] == "RW":
+                new_replica_ready = True
+                break
+        if new_replica_ready:
+            break
+        sleep(RETRY_INTERVAL)
+    assert new_replica_ready
+
+
+def wait_no_new_replica_ready(client, volume_name, replica_names):  # NOQA
+    """
+    Wait and make sure no new replicas are running on the specified
+    volume. Trigger a failed assertion of one is detected.
+    :param client: The Longhorn client to use in the request.
+    :param volume_name: The name of the volume.
+    :param replica_names: The list of names of the volume's old replicas.
+    """
+    new_replica_ready = False
+    for i in range(RETRY_COUNTS):
```

Comment:
This would be pretty long. Is there a way we can know for sure instead of waiting for the timeout?

Reply:
What `longhorn-ui` appears to be doing is checking for the field `volume.conditions.scheduled.status` and seeing if that's `true`. If that's `false`, then the frontend says that there's a scheduling error and handles it accordingly. We can probably check this field in the the test and, if necessary, combine this with checking the replica or some other volume attribute like the `Degraded state` or waiting some shorter period of time to ensure that the scheduling prevention is working properly.

I don't think we should go off of the replica by itself, because the replica can fail very quickly and get deleted, making it unreliable for us to check the scheduling condition.

Reply:
Yeah, something similar to `volume.conditions.scheduled.status` can help. But I am not sure if that condition works for this case as well. Since if scheduling failed, we shouldn't able to attach the volume.

Reply:
Wait, is that the expected behavior? 

It seems if you create a new `Volume` and there's no `Disks` available, there will be a scheduling error and you cannot attach. But when the `Volume` attempts to schedule an additional `Replica` while `Attached` but is failing to, then detaches, the `longhorn-manager` seems to attempt to stop rescheduling the new `Replica`. Even though the `numberOfReplicas` doesn't match the actual number of `Replicas` that exist, there's no attempt to schedule a new one, no scheduling error, and the `Volume` can still be attached again. Should we look into this?

Reply:
Sorry, I didn't make it clear. Updated my comment. The current behavior is correct. 

I see since we're not attempting scheduling a new one, so there is no error.

Reply:
Okay, this check has been updated to wait for a combination of `Degraded` state and scheduling failure to make sure that the `replica_scheduler` has refused to schedule a new `Replica` as it is supposed to.